### ISSUE TITLE
0.5 release prevents going back from Bytes to BytesMut

### ISSUE THREAD

**Issue by mqudsi** [tokio-rs/bytes#350]
Is the (new) inability to roundtrip between `Bytes` and `BytesMut` intentional? The inability to do so somewhat diminishes the usefulness of the reference counting feature this crate offers, if it means that  `freeze()` renders a chunk of memory permanently immutable with no way to reclaim borrowed ranges and start over again without reallocating.

I understand that `Bytes` is (now) unaware of the existence of `BytesMut` at this time, but it should still be possible to update the `BytesMut` api to recover memory  whose ref count has reached zero. I was all set to path `BytesMut` to add a version of `freeze()` that looked like this:

```rust
pub fn freeze() -> (Bytes, Unfreeze)
```

where `Unfreeze` is a struct with a single function 

```rust
pub fn unfreeze() -> BytesMut
```

with `Unfreeze` itself containing a weak reference to the same shared memory used by `Bytes` instances, but then I discovered that `Bytes` doesn't use `Arc` but rather a custom implementation that does not support weak references.

(The reason for having `BytesMut` return an `Unfreeze` rather than adding, e.g., an `unfreeze()` /`try_mut()` method to `Bytes` is that the latter would require `Bytes` keeping track of whether it was instantiated from read-only memory or not, which the current vtable mask isn't capable of distinguishing between.)

(My use case involves reading some bytes into a buffer, sending the ref-counted read bytes to another thread for processing, then reclaiming them when an event is set (and the `Bytes` have been dropped) to reuse the buffer for the next cycle.)

**Comment by seanmonstar** [comment]
Are you dropping the `BytesMut`, and wishing to get one back from the `Bytes` you now have? Or are you splitting off frozen `Bytes` while still retaining the `BytesMut`?

If the latter, you can just try to reserve space on it, and if all frozen `Bytes` have been dropped, the `BytesMut` can notice and reuse its memory.

**Comment by mqudsi** [comment]
Thanks for the quick reply.

> Are you dropping the `BytesMut`, and wishing to get one back from the `Bytes` you now have? Or are you splitting off frozen `Bytes` while still retaining the `BytesMut`?

I can structure it either way, but I was originally thinking the latter so that the original buffer size is a hard guarantee (to avoid all allocations).

> If the latter, you can just try to reserve space on it, and if all frozen `Bytes` have been dropped, the `BytesMut` can notice and reuse its memory.

With a possible exception of the case where the `BytesMut` write pointer is pointing to the very beginning of the remainder after a split, this would necessarily a) result in fragmentation of a once-contiguous underlying buffer (as bytes mapped to memory range 0..m now are being reused after range m..n) or else b) it would not be a zero-cost operation if the contiguous range is detected, the disjoint buffers unified, and the previous contents of the `BytesMut` post-split are then copied/moved to the beginning of the reclaimed range?

Even if fragmentation or lack of contiguity were not an issue, I think there would be merit in offering an explicit API, if only for a developer's peace of mind.

**Comment by andrewbaxter** [comment]
> If the latter, you can just try to reserve space on it, and if all frozen `Bytes` have been dropped, the `BytesMut` can notice and reuse its memory.

I stumbled across this from Tokio `Framed`, or at least I believe this is the right thread.

AFAICT `BytesMut` can't be reused because `freeze()` consumes the `BytesMut`.  I couldn't find another way to get a `Bytes` from `BytesMut`.

I'd expect to be able to do something like the following:
```
let mut my_bytes = BytesMut::with_capacity(999);
write_stuff(&mut my_bytes);
send(my_bytes.to_bytes()).await?;
my_bytes.clear();
write_stuff2(&mut my_bytes);
send(my_bytes.to_bytes()).await?;
...
```

I could do `.split().freeze()` but my understanding is that involves an allocation and copy too.

**Comment by Darksonn** [comment]
@andrewbaxter If the reference count is equal to one (i.e. all other split `Bytes`/`BytesMut` have been dropped) when the `BytesMut` tries to reallocate, then it will reuse the entire allocation even if you have previously done a `split`+`freeze`.

**Comment by andrewbaxter** [comment]
Ah, thanks for the quick reply!  Okay, so that usage is idiomatic.

I think my confusion arose basically from crate documentation that says `This is managed by using a reference count to track when the memory is no longer needed and can be freed.` wrt Bytes, but doesn't mention writing or `BytesMut` and I would expect those to have different constraints.  Since `Bytes` is RO, I thought the reference counting was for zero copy parsing, rather than reusing write buffers.

Reading deeper into the `BytesMut` documentation I guess when `Bytes` is obtained from `BytesMut` both `Bytes` _and_ `BytesMut` are views of a single underlying data store, and the reference count in `Bytes` in this case is actually the reference count on the data store shared by `BytesMut` rather than a `Bytes`-specific reference count.

So this would be the ideal usage for write/read/reuse:
```
let mut my_bytes = BytesMut::with_capacity(999);
write_stuff(&mut my_bytes);
send(my_bytes.split().freeze()).await?;
write_stuff2(&mut my_bytes);
send(my_bytes.split().freeze()).await?;
```

**Comment by Darksonn** [comment]
Indeed, a single allocation can contain any number of `BytesMut` and `Bytes` pointing into it as long as each `BytesMut` does not overlap with the range of any other handle.

**Comment by Lucretiel** [comment]
> This is available as `BytesMut::from(bytes)`

What happens if the `Bytes` *aren't* a unique view of the memory? Shoud this be `try_from`?

EDIT: Ah, I see, it's basically `Cow::to_mut`, where it clones if the original view isn't unique. Makes sense to me.
